Libs/Mesh/Mesh.cpp

```diff
@@ -967,6 +989,599 @@ std::ostream& operator<<(std::ostream &os, const Mesh& mesh)
+
+        vtkSmartPointer<vtkSelectPolyData> select = vtkSelectPolyData::New();
+        select->SetLoop(selectionPoints);
+        select->SetInputData(this->mesh);
+        select->GenerateSelectionScalarsOn();
+        select->SetSelectionModeToLargestRegion();
+
+        // Clipping mesh
+        vtkSmartPointer<vtkClipPolyData> selectclip = vtkClipPolyData::New();
+        selectclip->SetInputConnection(select->GetOutputPort());
+        selectclip->SetValue(0.0);
+
+        selectclip->Update();
+
+        MeshType halfmesh = selectclip->GetOutput();
+
+        vtkSmartPointer<vtkDoubleArray> inout = computeInOutForFFCs(query, halfmesh);
+
+
+
+        vtkSmartPointer<vtkDoubleArray> values = vtkSmartPointer<vtkDoubleArray>::New();
+        vtkSmartPointer<vtkDoubleArray> absvalues = setDistanceToBoundaryValueFieldForFFCs(values, points, boundaryVerts, inout, V, F, dom);
+
+        this->mesh->GetPointData()->SetActiveScalars("value");
+
+        std::vector<Eigen::Matrix3d> face_grad_ = this->setGradientFieldForFFCs(absvalues, V, F);
+
+        //this->visualizeVectorFieldForFFCs(values, face_grad_, V, F);
+
+    } // Per boundary for loop end
+
+    locator = vtkSmartPointer<vtkCellLocator>::New();
+    locator->SetDataSet(this->mesh);
+    locator->BuildLocator();
+
+    // Write mesh for debug purposes
+//    std::string fnin = "dev/mesh_" + std::to_string(dom) + "_" + std::to_string(num) + "_in.vtk";
+//    this->write(fnin);
+
+     return true;
+}
+
+// Copied directly from Meshwrapper
+Eigen::Vector3d Mesh::ComputeBarycentricCoordinates(const Eigen::Vector3d& pt, int face) const
+{
+  double closest[3];
+  int sub_id;
+  double pcoords[3];
+  double dist2;
+  Eigen::Vector3d bary;
+  this->mesh->GetCell(face)->EvaluatePosition(pt.data(), closest, sub_id, pcoords, dist2, bary.data());
+  return bary;
+}
+
+double Mesh::GetFFCValue(Eigen::Vector3d query){
+    locator->BuildLocator();
+
+    double closestPoint[3];
+    vtkIdType cellId;
+    int subId;
+    double dist;
+    locator->FindClosestPoint(query.data(), closestPoint, cellId, subId, dist);
+
+    auto cell = this->mesh->GetCell(cellId);
+
+    size_t v1 = cell->GetPointId(0); size_t v2 = cell->GetPointId(1); size_t v3 = cell->GetPointId(2);
+
+    // Compute baricentric distances
+    Eigen::Vector3d cp(closestPoint[0], closestPoint[1], closestPoint[2]);
+    Eigen::Vector3d bary = ComputeBarycentricCoordinates(cp, cellId);
+
+    bary = bary/bary.sum();
+
+    Eigen::Vector3d values(this->getFieldValue("value", v1), this->getFieldValue("value", v2), this->getFieldValue("value", v3));
+
+    return (bary*values.transpose()).mean();
+}
+
+Eigen::Vector3d Mesh::GetFFCGradient(Eigen::Vector3d query){
+
+    locator->BuildLocator();
+
+    double closestPoint[3];
+    vtkIdType cellId;
+    int subId;
+    double dist;
+    locator->FindClosestPoint(query.data(), closestPoint, cellId, subId, dist);
```

Comment:
Have you tried using `Mesh.closestPointId` or `Mesh.closestPoint`?

Reply:
These functions create a locator every time they run. I don't think it's very efficient for me to use these since I query this a lot. I also don't think a locator should be created for these functions, since they need to be general enough.